zkvm/src/signature/prover.rs

```diff
@@ -0,0 +1,187 @@
+    x_i: Scalar,
+    r_i: Scalar,
+    R_i: NonceCommitment,
+}
+
+pub struct PartyAwaitingCommitments {
+    transcript: Transcript,
+    multikey: Multikey,
+    x_i: Scalar,
+    r_i: Scalar,
+    nonce_precommitments: Vec<NoncePrecommitment>,
+}
+
+pub struct PartyAwaitingSiglets {
+    multikey: Multikey,
+    c: Scalar,
+    R: RistrettoPoint,
+    nonce_commitments: Vec<NonceCommitment>,
+}
+
+impl NonceCommitment {
+    fn precommit(&self) -> NoncePrecommitment {
+        let mut h = Transcript::new(b"MuSig.nonce-precommit");
+        h.commit_point(b"R", &self.0.compress());
+        let mut precommitment = [0u8; 32];
+        h.challenge_bytes(b"precommitment", &mut precommitment);
+        NoncePrecommitment(precommitment)
+    }
+}
+
+impl PartyAwaitingPrecommitments {
+    pub fn new(
+        // The message `m` should already have been fed into the transcript
+        transcript: &Transcript,
+        x_i: Scalar,
+        multikey: Multikey,
+    ) -> (Self, NoncePrecommitment) {
+        let mut rng = transcript
+            .build_rng()
+            .commit_witness_bytes(b"x_i", &x_i.to_bytes())
+            .finalize(&mut rand::thread_rng());
+
+        // Generate ephemeral keypair (r_i, R_i). r_i is a random nonce.
+        let r_i = Scalar::random(&mut rng);
+        // R_i = generator * r_i
+        let R_i = NonceCommitment(RISTRETTO_BASEPOINT_POINT * r_i);
+        // Make H(R_i)
+        let precommitment = R_i.precommit();
+
+        (
+            PartyAwaitingPrecommitments {
+                transcript: transcript.clone(),
+                multikey,
+                x_i,
+                r_i,
+                R_i,
+            },
+            precommitment,
+        )
+    }
+
+    pub fn receive_precommitments(
+        self,
+        nonce_precommitments: Vec<NoncePrecommitment>,
+    ) -> (PartyAwaitingCommitments, NonceCommitment) {
+        // Store received nonce precommitments in next state
+        (
+            PartyAwaitingCommitments {
+                nonce_precommitments,
+                transcript: self.transcript,
+                multikey: self.multikey,
+                x_i: self.x_i,
+                r_i: self.r_i,
+            },
+            self.R_i,
+        )
+    }
+}
+
+impl PartyAwaitingCommitments {
+    pub fn receive_commitments(
+        mut self,
+        nonce_commitments: Vec<NonceCommitment>,
+    ) -> Result<(PartyAwaitingSiglets, Scalar), VMError> {
+        // Check stored precommitments against received commitments
+        for (pre_comm, comm) in self
+            .nonce_precommitments
+            .iter()
+            .zip(nonce_commitments.iter())
+        {
+            // Make H(comm) = H(R_i)
+            let actual_precomm = comm.precommit();
+
+            // Compare H(comm) with pre_comm, they should be equal
+            let equal = pre_comm.0.ct_eq(&actual_precomm.0);
+            if equal.unwrap_u8() == 0 {
+                return Err(VMError::InconsistentWitness);
```

Comment:
would be good to signal which party is at fault, so we can decrease their karma

Reply:
the error must be musig-specific, like `MuSigError::InvalidNonceCommitment(usize)`, similarly for share validation.

Reply:
good call!